File: src/regwatch/common/ratelimit.py
"""In-memory, per-process rate limiting.

Pilot scope: one API process, so a threading.Lock plus a deque of monotonic
timestamps per key is sufficient. Distributed rate limiting belongs to the IT
gateway in production (docs/PROD_READINESS.md #1).
"""

from __future__ import annotations

import threading
import time
from collections import deque

# NOTE (scope): this limiter is in-process, so under fly.toml's
# min_machines_running=2 each machine keeps its own window and the EFFECTIVE
# ceiling is ~2x the configured limit. A shared-store (Redis/Postgres) limiter
# is the fix for an exact global cap, but that is a separate, parked item
# (docs/PROD_READINESS.md #1) - do NOT build it here. The LOGIN limiter moved
# to the Go proxy with the step-4 auth cutover (go/internal/api/ratelimit.go
# ports this class); only query_limiter remains Python-side.


class RateLimiter:
    """Sliding one-minute window per key. ``limit <= 0`` disables the check."""

    def __init__(self, window_s: float = 60.0) -> None:
        self._window_s = window_s
        self._lock = threading.Lock()
        self._hits: dict[str, deque[float]] = {}
        self._last_sweep = 0.0

    def allow(self, key: str, limit: int) -> bool:
        if limit <= 0:
            return True
        now = time.monotonic()
        with self._lock:
            self._sweep(now)
            hits = self._hits.setdefault(key, deque())
            while hits and now - hits[0] >= self._window_s:
                hits.popleft()
            if len(hits) >= limit:
                return False
            hits.append(now)
            return True

    def _sweep(self, now: float) -> None:
        """Evict keys whose whole window has expired. Caller holds the lock.

        Keys can embed caller-supplied identifiers, so without eviction a
        spray of unique keys grows the dict for the life of the process (the
        original motivation was login emails, now the Go limiter's problem;
        the guard stays for any future key shape). One O(keys) pass per window.
        """
        if now - self._last_sweep < self._window_s:
            return
        self._last_sweep = now
        stale = [
            key for key, hits in self._hits.items() if not hits or now - hits[-1] >= self._window_s
        ]
        for key in stale:
            del self._hits[key]

    def reset(self) -> None:
        with self._lock:
            self._hits.clear()


query_limiter = RateLimiter()


def reset_for_tests() -> None:
    """Clear limiter state so one test's traffic cannot 429 the next."""
    query_limiter.reset()

File: go/internal/api/ratelimit.go
package api

import (
	"sync"
	"time"
)

// Port of src/regwatch/common/ratelimit.py -- in-memory, per-process sliding
// one-minute window. The caps are FIXED, not settings (the Python file: "there
// is no legitimate reason to raise it"), and the limiter being per-process is
// the SAME scope as Python's: under min_machines_running=2 each machine keeps
// its own window and the effective ceiling is ~2x these numbers, exactly as
// documented there. A shared-store limiter stays a parked item.
const (
	// Login brute-force guard: attempts per email per minute.
	LoginAttemptsPerMinute = 10
	// Companion per-IP cap so spraying many DISTINCT emails from one host is
	// also throttled; above the per-email cap so a shared office/VPN NAT is
	// not locked out at the first wrong password.
	LoginAttemptsPerIPPerMinute = 30

	rateWindow = time.Minute
)

// RateLimiter is a sliding one-minute window per key. limit <= 0 disables the
// check. now is injectable for tests (Python uses time.monotonic; Go's
// time.Time carries a monotonic clock on time.Now()).
type RateLimiter struct {
	mu        sync.Mutex
	hits      map[string][]time.Time
	lastSweep time.Time
	now       func() time.Time
}

func NewRateLimiter() *RateLimiter {
	return &RateLimiter{hits: map[string][]time.Time{}, now: time.Now}
}

// Allow reports whether key may take another hit, consuming a slot when it
// may. Mirrors ratelimit.py::RateLimiter.allow exactly, including the
// consume-on-success behavior (successful logins also charge the bucket).
func (l *RateLimiter) Allow(key string, limit int) bool {
	if limit <= 0 {
		return true
	}
	now := l.now()
	l.mu.Lock()
	defer l.mu.Unlock()
	l.sweep(now)
	hits := l.hits[key]
	for len(hits) > 0 && now.Sub(hits[0]) >= rateWindow {
		hits = hits[1:]
	}
	if len(hits) >= limit {
		l.hits[key] = hits
		return false
	}
	l.hits[key] = append(hits, now)
	return true
}

// sweep evicts keys whose whole window has expired; caller holds the lock.
// Login keys embed the client-supplied email, so without eviction an
// unauthenticated caller spraying unique emails grows the map for the life of
// the process. At most one O(keys) pass per window.
func (l *RateLimiter) sweep(now time.Time) {
	if now.Sub(l.lastSweep) < rateWindow {
		return
	}
	l.lastSweep = now
	for key, hits := range l.hits {
		if len(hits) == 0 || now.Sub(hits[len(hits)-1]) >= rateWindow {
			delete(l.hits, key)
		}
	}
}

// Reset clears limiter state (the reset_for_tests parity hook).
func (l *RateLimiter) Reset() {
	l.mu.Lock()
	defer l.mu.Unlock()
	l.hits = map[string][]time.Time{}
}
